sdk/include/opentelemetry/sdk/common/attribute_utils.h

```diff
@@ -33,22 +33,44 @@ namespace common
  */
 using OwnedAttributeValue = nostd::variant<bool,
                                            int32_t,
-                                           uint32_t,
                                            int64_t,
+                                           uint32_t,
                                            uint64_t,
                                            double,
                                            std::string,
+// TODO: should we have support for C-strings here or std::string is sufficient?
```

Comment:
In this exact case - for the owning class, I agree that we probably don't want the C-strings.. I can remove this comment.

For the outside containers on API surface, I see some value to have both - `std::string` and `const char *` / string literal initialization. Also for the gRPC / OTLP, - protobuf itself accept both `std::string` and `const char *` as arguments during serialization. It would be reasonable to support both on API surface.

Reply:
I think from the API point of view it shouldn't matter that much. `const char*` will be automatically converted to `std::string`:
```c++
std::variant<std::string> x;

x = "abc";
x = std::string("def");
```
I don't have too strong an opinion here. Mostly I'm in favor of seeing the types that the variants hold being kept to a mininum, as otherwise it makes it harder to consume a variant (checking all possible types), which is a pain anyway.

Reply:
`string_view` itself is more of a `const char *` rather than `std::string`.

The idea is to support `const char *` upfront for future-proofing and to remain compatible with C API, if one to be ever built on top of C++ API, without needing a memcpy.

In essence - you do not necessarily need to construct a copy in `std::string` object every time, if you know the buffer & length, and if you can process the original `Event` without copying it. This also helps to easily build a C API on top of C++ API --- because you can still express an array of attributes in C, for example, and a for each key-value iterator on top of that array. For example, using a structure like here:
https://github.com/microsoft/cpp_client_telemetry/blob/3d48e213cb8e4a5693042f06ae7676901d05f560/lib/include/public/mat.h#L167

```cpp
    typedef union
    {
        /* Basic types */
        uint64_t            as_uint64;
        const char*         as_string;
        int64_t             as_int64;
        double              as_double;
        bool                as_bool;
        evt_guid_t*         as_guid;
        uint64_t            as_time;
        /* Array types are nullptr-terminated array of pointers */
        char**              as_arr_string;
        int64_t**           as_arr_int64;
        bool**              as_arr_bool;
        double**            as_arr_double;
        evt_guid_t**        as_arr_guid;
        uint64_t**          as_arr_time;
    } evt_prop_v;
```

Demanding this owned value to be always `std::string` -- implies a `memcpy` , that could be otherwise avoided. I believe this is also the reason why gRPC itself accepts a string to be either `std::string` or `const char *`, and I added the necessary fix for that here before:
https://github.com/open-telemetry/opentelemetry-cpp/blob/4751ae5a29e5eeed1b775b7c337bdbe294165346/exporters/otlp/src/recordable.cc#L65

From spec perspective, both are `strings` (on-wire), but from API perspective - both could be either this or that string, that should be totally within the constraints of the API spec. There is also `std::wstring`  😄 , but I'd rather not go there -- instead I'll offer an option how to transform that to one of the supported types in the updated iteration.